File: benchmarks/fixtures/webapp_ts/src/cache/redisCache.ts
/** Redis-backed cache. */
import { getLogger } from '../utils/helpers';
import { BaseCache } from './base';

const logger = getLogger("cache.redis");

/** Redis cache implementation. */
export class RedisCache extends BaseCache {
    private store: Map<string, unknown> = new Map();
    private expiry: Map<string, number> = new Map();

    constructor(host: string = "localhost", port: number = 6379) {
        super("redis");
        logger.info(`RedisCache created: ${host}:${port}`);
    }

    get(key: string): unknown | null {
        if (this.store.has(key)) {
            const exp = this.expiry.get(key) ?? Infinity;
            if (Date.now() > exp) {
                this.store.delete(key);
                this.expiry.delete(key);
                this.misses++;
                return null;
            }
            this.hits++;
            return this.store.get(key) ?? null;
        }
        this.misses++;
        return null;
    }

    set(key: string, value: unknown, ttl: number = 300): void {
        this.store.set(key, value);
        this.expiry.set(key, Date.now() + ttl * 1000);
        logger.info(`Redis SET ${key} (ttl=${ttl})`);
    }

    delete(key: string): boolean {
        this.expiry.delete(key);
        return this.store.delete(key);
    }

    clear(): number {
        const count = this.store.size;
        this.store.clear();
        this.expiry.clear();
        logger.info(`Redis FLUSHDB: ${count} keys`);
        return count;
    }

    incr(key: string, amount: number = 1): number {
        const current = (this.store.get(key) as number) ?? 0;
        const newVal = current + amount;
        this.store.set(key, newVal);
        return newVal;
    }
}

File: benchmarks/fixtures/webapp_rb/cache/redis_cache.rb
# Redis-backed cache.

require_relative '../utils/helpers'
require_relative 'base_cache'

# Redis cache implementation.
class RedisCache < BaseCache
  def initialize(host = 'localhost', port = 6379)
    super('redis')
    @store = {}
    @expiry = {}
    get_logger('cache.redis').info("RedisCache created: #{host}:#{port}")
  end

  def get(key)
    if @store.key?(key)
      exp = @expiry[key] || Float::INFINITY
      if Time.now.to_i > exp
        @store.delete(key)
        @expiry.delete(key)
        @misses += 1
        return nil
      end
      @hits += 1
      return @store[key]
    end
    @misses += 1
    nil
  end

  def set(key, value, ttl = 300)
    @store[key] = value
    @expiry[key] = Time.now.to_i + ttl
    get_logger('cache.redis').info("Redis SET #{key} (ttl=#{ttl})")
  end

  def delete(key)
    @expiry.delete(key)
    !@store.delete(key).nil?
  end

  def clear
    count = @store.size
    @store.clear
    @expiry.clear
    get_logger('cache.redis').info("Redis FLUSHDB: #{count} keys")
    count
  end

  def incr(key, amount = 1)
    current = @store[key] || 0
    new_val = current + amount
    @store[key] = new_val
    new_val
  end
end
